### PR TITLE
signrpc: Add no_hashing flag to signMessage

### LINKED ISSUE #4474: Allow to sign a message directly
### Background

I want to implement LNURL-auth.
For it to work a challenge (k1) has to be signed.
The service where you want to authenticate verifies that you have provided a valid signature and logs you in if it was valid.

The problem is that LNURL-auth spec requires to verify the signature for k1, while LND creates the signature for sha256(k1).
Therefore the signature validation fails and it is currently not possible to implement LNURL-auth with LND.

Can we have an additional boolean parameter for the sign command to specify whether to sign the hash of the message or the message directly? And have this exposed to gRPC?

For LNURL-auth specs please see here:
https://github.com/btcontract/lnurl-rfc/blob/master/lnurl-auth.md

### PR DESCRIPTION
**Pull request by hsjoberg**
Closes #4474.

To be able to support [lnurl-auth](https://github.com/btcontract/lnurl-rfc/blob/master/lnurl-auth.md), a message ("k1") needs to be signed by lnd without it being hashed.
This PR addresses this by adding a new flag `no_hashing` to the `signrpc.signMessage` method. It defaults to false to keep current behavior. If it is set to true, no hashing of the message will be made before signing.

I have not updated `verifyMessage` with the flag but I could do that as well if you want me to.

Cheers

### CODE REVIEW

File: keychain/btcwallet.go [hunk 1]
```diff
@@ -379,13 +379,13 @@ func (b *BtcWalletKeyRing) ECDH(keyDesc KeyDescriptor,
 //
 // NOTE: This is part of the keychain.DigestSignerRing interface.
 func (b *BtcWalletKeyRing) SignDigest(keyDesc KeyDescriptor,
-	digest [32]byte) (*btcec.Signature, error) {
+	digest []byte) (*btcec.Signature, error) {
```
Comment (yyforyongyu) [inline keychain/btcwallet.go:382]: The description above needs to be updated if the function were to take arbitrary bytes. However, `SignDigest`, as the name indicates, is signing a `digest`, so I would expect the `digest` parameter to be a hash digest, not arbitrary length of bytes. Plus, the `privKey.Sign` is using the underlying `signRFC6979`, which expects a hash digest.

File: lnrpc/signrpc/signer_server.go [hunk 2]
```diff
@@ -446,13 +446,17 @@ func (s *Server) SignMessage(ctx context.Context,
 		},
 	}
 
-	// The signature is over the sha256 hash of the message.
 	var digest [32]byte
-	copy(digest[:], chainhash.HashB(in.Msg))
+	if in.RawMsg {
+		copy(digest[:], in.Msg)
```
Comment (yyforyongyu) [inline lnrpc/signrpc/signer_server.go:451]: It looks like the digest is 32 byte anyway, so I guess we could leave the `SignDigest` stay untouched?

File: lnrpc/signrpc/signer_server.go [hunk 3]
```diff
@@ -446,13 +446,17 @@ func (s *Server) SignMessage(ctx context.Context,
...
+	} else {
+		// The signature is by default over the sha256 hash of the message.
```
Comment (yyforyongyu) [inline lnrpc/signrpc/signer_server.go:453]: nit: line too long

**Comment by hsjoberg** [comment]
Thank you for your feedback @yyforyongyu!  
I didn't know that cryptographic signing requires a fixed amount of bytes, but yes that absolutely makes sense.

I made the code prior to this knowledge, which is why I named the flag to the rather cryptic `raw_msg`.
Perhaps naming it `no_hashing` or something similar and mandating the input to be 32 bytes makes more sense.
This is also why removed the 32 bytes declaration to various functions (go failed to compile because `msg` is of type `[]byte`).

File: keychain/btcwallet.go (hunk 1 above)
Comment (hsjoberg) [reply keychain/btcwallet.go]: Makes sense. Fixed by reverting back to `[32]byte`.

File: lnrpc/signrpc/signer_server.go (hunk 3 above)
Comment (hsjoberg) [reply lnrpc/signrpc/signer_server.go:456]: It's actually within the 80 column rule. 
image
I'm not sure if it's a good idea to split it to two rows.

File: lnrpc/signrpc/signer_server.go (hunk 2 above)
Comment (hsjoberg) [reply lnrpc/signrpc/signer_server.go:454]: You're right, fixed.

**Comment by hsjoberg** [comment]
@yyforyongyu Please take another look, I think I addressed all your points.

File: lnrpc/signrpc/signer_server.go (hunk 3 above)
Comment (guggero) [reply lnrpc/signrpc/signer_server.go:456]: Not if you use 8 spaces for the tab character, which is the default for golang and described in our [contribution guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#61-contribution-checklist).

Comment (hsjoberg) [reply lnrpc/signrpc/signer_server.go]: @guggero thanks! I missed this.
Fixed.

**Comment by hsjoberg** [comment]
Rebased and capitalized "MUST" in error message (`msg MUST be 32 bytes`).

**Comment by guggero** [comment]
Thank you for the PR.
Unfortunately I have serious security concerns with this seemingly simple change. I see why this is wanted, but let me first ask some questions (also directed at @michaelWuensch and @btccontract):

1. Why would the `lnurl-auth` scheme not require `k1` to be hashed before signing it? This is even a value that is user provided. I think this would make it quite trivial for an attacker to extract private key material with just a few calls with specially constructed `k1` values.
2. You cannot sign with a key under `m/138'/0` with this RPC anyway. How would you get around that obstacle? Or would there be a special "lnd lnurl auth scheme"?

While I do think the intentions (and the need for) `lnurl-auth` is good and justified, I worry that it hasn't been looked at properly from a cryptography standpoint. "Rolling your own crypto protocol" just never is a good idea.

@cfromknecht what's your opinion here?

**Comment by hsjoberg** [comment]
Hi @guggero, I appreciate your reply.

> 1. Why would the lnurl-auth scheme not require k1 to be hashed before signing it? This is even a value that is user provided. I think this would make it quite trivial for an attacker to extract private key material with just a few calls with specially constructed k1 values.

I think this question is best suited for @btcontract.  
It [looks](https://github.com/btcontract/lnurl-rfc/blob/master/lnurl-auth.md) like this could be changed so that the wallet will always hash the message before signing it.

Unfortunately the current schema is widely deployed by different [wallets](https://github.com/fiatjaf/awesome-lnurl#wallets) and [services](https://github.com/fiatjaf/awesome-lnurl#lnurl-auth). But if this is a security concern, I think it definitely makes sense to address this.

> 2. You cannot sign with a key under m/138'/0 with this RPC anyway. How would you get around that obstacle? Or would there be a special "lnd lnurl auth scheme"?

Yes, it's not possible to get the `hashingKey` according to the specs, as we cannot do hmacSHA256 on the private key derived from `m/138'/0` using the RPC.  
It's also not possible to use this `hashingKey` to be fully complaint on how to get the `linkingKey`, because lnd doesn't support full [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki) derivation, only two levels deep ([BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)).

So there will have to be special cases for the `linkingKey` and `hashingKey` with lnd.

**Comment by cfromknecht** [comment]
@guggero your spidey senses are not unfounded.

> I think this would make it quite trivial for an attacker to extract private key material with just a few calls with specially constructed k1 values.

No, this does not leak your private key.

That said, opening this PR did lead to the identification of weaknesses in the construction. The relevant details were disclosed privately to @btcontract 2-3 weeks ago and we remain in contact. Even though there are inherent weaknesses in the primitives, **there doesn't appear to be any immediate security implications for LNURL as specified**. However, we both agree that these weaknesses are worth fixing and are actively investigating a v2 scheme for lnurl-auth. 

@hsjoberg As it relates to this PR, unfortunately I am not comfortable exposing insecure signing primitives generically via lnd's rpc interface. Therefore we probably won't see lnurl-auth support in lnd until a replacement is standardized, however we are working to make that happen sooner than later.

NOTE: I consulted @btcontract before replying to ensure we were both in agreed to this preliminary disclosure given the security risk is minimal. More details will be available soon, as of now there is no immediate action that needs to be taken by clients or servers.

**Comment by hsjoberg** [comment]
> @hsjoberg As it relates to this PR, unfortunately I am not comfortable exposing insecure signing primitives generically via lnd's rpc interface.

@cfromknecht Totally understandable. As long as we come up with another solution that works for everyone, I'm happy.

> Therefore we probably won't see lnurl-auth support in lnd until a replacement is standardized, however we are working to make that happen sooner than later.

Sounds good, keep in the loop if there's anything I can do.

Cheers

**Pull request closed without merging**